jujugui/static/gui/src/app/init/app.js

```diff
@@ -0,0 +1,1465 @@
+    // Set up a handler for syncing the service token hover. This needs to be
+    // attached only when the component is visible otherwise the added
+    // services component will try to render if the user hovers a service
+    // when they have the service details open.
+    const hoverHandler = this._domEventHandlers['topo.hoverService'];
+    if (hoverHandler) {
+      document.removeEventListener('topo.hoverService', hoverHandler);
+    }
+    this._domEventHandlers['topo.hoverService'] = evt => {
+      this.setState({ hoveredService: evt.detail.id });
+    };
+    document.addEventListener(
+      'topo.hoverService', this._domEventHandlers['topo.hoverService']);
+    // Deselect the active service token. This needs to happen so that when a
+    // user closes the service details the service token deactivates.
+    ServiceModule.deselectNodes();
+    const db = this.props.db;
+    return (
+      <Panel
+        instanceName="inspector-panel"
+        visible={db.services.size() > 0}>
+        <AddedServicesList
+          changeState={this._bound.changeState}
+          hoveredId={this.state.hoveredService}
+          serviceModule={
+            shapeup.fromShape(ServiceModule, AddedServicesList.propTypes.serviceModule)}
+          services={db.services} />
+      </Panel>);
+  }
+  /**
+    Renders the Environment Size Display component to the page in the
+    designated element.
+  */
+  _generateEnvSizeDisplay() {
+    const { db } = this.props;
+    return (
+      <EnvSizeDisplay
+        appState={this.props.appState}
+        machineCount={db.machines.filterByParent().length}
+        providerType={this.props.modelAPI.get('providerType') || ''}
+        serviceCount={db.services.size()} />);
+  }
+  /**
+    Renders the model action components to the page in the designated
+    element.
+  */
+  _generateModelActions() {
+    const modelAPI = this.props.modelAPI;
+    return (
+      <div id="model-actions-container">
+        <ModelActions
+          acl={this.props.acl}
+          appState={this.props.appState}
+          changeState={this._bound.changeState}
+          displayTerminalButton={this._enableTerminal()}
+          exportEnvironmentFile={
+            initUtils.exportEnvironmentFile.bind(initUtils, this.props.db)}
+          hideDragOverNotification={this._showDragOverNotification.bind(this, false)}
+          importBundleFile={this.props.bundleImporter.importBundleFile.bind(
+            this.props.bundleImporter)}
+          loadingModel={modelAPI.loading}
+          renderDragOverNotification={
+            this._showDragOverNotification.bind(this)}
+          sharingVisibility={this._sharingVisibility.bind(this)}
+          userIsAuthenticated={modelAPI.userIsAuthenticated} />
+      </div>);
+  }
+  /**
+    Display or hide the sharing modal.
+    @param {Boolean} visibility Controls whether to show (true) or hide
+                     (false); defaults to true.
+  */
+  _sharingVisibility(visibility = true) {
```

Comment:
It might be helpful to rename the state lib to URLState or something. I thought about moving the App component state into the state lib but then it's hard to know what should be URL state vs temp UI state.